dashboard/app/controllers/regional_partners_controller.rb

```diff
@@ -152,4 +186,79 @@ def regional_partner_params
   def restricted_users
     User.select(RESTRICTED_USER_ATTRIBUTES_FOR_VIEW)
   end
+
+  # Validate that file for bulk replace of partner mappings exists,
+  # has a 'Region' header and is a valid CSV file. If any of the above
+  # are false, returns false and sets an alert instead of returning CSV
+  # as an object
+  def validate_mappings_csv(file)
+    unless file
+      flash[:alert] = "Replace mappings failed. CSV file not found."
+      return false
+    end
+    begin
+      csv = CSV.read(file.path, headers: true, liberal_parsing: true)
+      unless csv.headers.include?("Region")
+        flash[:alert] = "Replace mappings failed. CSV does not include 'Region' header."
+        return false
+      end
+      return csv
+    rescue
+      flash[:alert] = "Replace mappings failed. Could not read file."
+      return false
+    end
+  end
+
+  # Given a CSV object of region mappings, build and validate each mapping
+  # and, if valid, add to mappings array. If it is invalid add an
+  # error to the errors array in the format {region, message}
+  # Mappings are not added to any regional partner in this method.
+  def parse_mappings_from_csv(csv, mappings, errors)
+    regions_missing = false
+    csv.each do |row|
+      region = row['Region']
+      unless region.present?
+        regions_missing = true
+        next
+      end
+      state = region if region.in?(STATE_ABBR_WITH_DC_HASH.keys.map(&:to_s))
+      zip_code = region if RegexpUtils.us_zip_code?(region)
+      if !state && !zip_code
+        errors << {region: region, message: "Invalid region"}
+        next
+      end
+      mapping = @regional_partner.mappings.build({state: state, zip_code: zip_code})
+      mapping.valid?
+      if !mapping.errors[:base].nil_or_empty?
```

Comment:
Could these two lines just be
```rb
unless mapping.valid?
```
How important is it that we check `errors[:base]` specifically?

Reply:
I think you are right, fixed